### ISSUE TITLE
Memory Leak with MultiAgentEnvironment

### ISSUE THREAD

**Issue by RolandZimmermann** [ray-project/ray#6359]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 18.04.3 LTS
- **Ray installed from (source or binary)**: pip installation
- **Ray version**: 0.7.5
- **Python version**: 3.6.8
- **Exact command to reproduce**: -

### Describe the problem

We are running a multiagent environment with our own environment and custom models for the agents. There are multiple problems occuring during the training with PPO. Our system has 16 cores and 32 GB of ram. The trainer reports that only 1 out of 16 CPUs is used, but the num_of_healty_workers reported is equal to 15. Also, all cores are under workload. Why is it showing only 1/16? This doesn't occure when we run the Cartpole environment. There it correctly shows 3/16 while using 2 cores as worker.

![image](https://user-images.githubusercontent.com/37594534/70137246-d2241000-168d-11ea-9c1e-f7ea5782e422.png)

There also seems to be a problem with the amount of memory usage. We specified to only use 30 GB but it always defaults to 32GB as in the picture above. Setting different sizes for object_memory and redis_memory doesn't seem to have an effect.

The memory usage is rising an unknown amount of time, then it slows down but still rises until the memory runs out.

![image](https://user-images.githubusercontent.com/37594534/70137340-fb44a080-168d-11ea-88d2-f22990fbc741.png)

![image](https://user-images.githubusercontent.com/37594534/70137357-03044500-168e-11ea-9436-b743487ff807.png)

Increasing the number of agents causes the system to crash even faster, way beyond 32GB or even 64GB.

Our custom environment is not causing the memory leak. We can run multiple instances of the environment with constant memory usage.

Also, if we try to run a grid search with different learning rates it gets stuck at the beginning. After some time it reports: [RayletClient] Connection closed unexpectedly.

![image](https://user-images.githubusercontent.com/37594534/70137390-157e7e80-168e-11ea-8d85-ecff295d1a97.png)

### Source code / logs

Our config for the trainer looks as following:
```
# reinforcement learning
# related options
rl:

    # config that gets passed to the trainer,
    # e.g. PPOTrainer
    trainer:

        # model spec
        model:
            custom_model: "our_model"
            custom_options: 
                evaluation: False
                restore: False
            use_lstm: False
            fcnet_hiddens: [128, 1024]

        # resources
        num_gpus: 0
        num_envs_per_worker: 1

        # logging
        log_level: "INFO"

        # disable filters
        observation_filter: "NoFilter"

        # dont use a gpu (?)
        simple_optimizer: True

        # share layers between
        # value function and policy model
        vf_share_layers: False

        # sample sizes
        sample_batch_size: 2000 
        sgd_minibatch_size: 128 

        # steps the simulation has to run for until training kicks in
        train_batch_size: 40000 
```

The config for ray looks as following:

```
# ray config options
ray:
    webui: false

    redis_port: 6379
    redis_addr: "localhost:6379"

    checkpoint_file: "checkpoint.out"

    # number of cpus to use for ray / training
    num_cpus: 16

    # available ram in gigabytes
    ram: 30.0
```

The way we start ray:
```
# initialize ray
# allocate half of memory for object store
fraction_of_ram = int(ram_avail_bytes // 5)
ray.init(
    num_cpus=cpus,
    redis_max_memory=fraction_of_ram,
    object_store_memory=fraction_of_ram,
)

# register environment
register_env(env_name, lambda ec: SimulationEnv(ec))

# go
tune.run(
   train,
   name=exp_name,
   config={
   "env_config": {"config": cfg},
   "num_workers": cpus - 1,
    **trainer_conf,
    },
)
```

**Comment by ericl** [comment]
Re: 1/16 CPUs, this has to do with how you're using a custom "train" func. When you do this, you need to specify the resources explicitly in the tune config (with the built in trainers, RLlib does this automatically). So it's actually using 16 CPUs but Tune just "thinks" the trial is using 1.

For memory -- I would recommend not setting redis max memory and object store memory so high. Instead, I would `ray.init(object_store_memory=1e9)` or something like that to constrain object memory (RLlib generally does not use huge amounts of shared memory unless you have huge batches of image observations or things like that).

**Comment by eugenevinitsky** [comment]
Experiencing the same thing on the nightly builds despite setting ray.init(object_store_memory=int(5e9), memory=int(5e9), redis_max_memory=int(1e9))

**Comment by ericl** [comment]
Is this a regression? Would be great to have a repro script that uses a dummy env for debugging.

**Comment by eugenevinitsky** [comment]
Yeah, let me try and put one up ASAP

**Comment by eugenevinitsky** [comment]
I can't reproduce on my local machine but I have an ec2 instance (I can send you the key) where the following script slowly has this behavior. 

`from gym import Env
from gym.spaces import Box

import numpy as np

import ray
import ray.rllib.agents.ppo as ppo
from ray import tune
from ray.tune import run as run_tune
from ray.tune.registry import register_env

class DummyEnv(Env):

    @property
    def observation_space(self):
        return Box(low=0, high=1, shape=(20, ))

    @property
    def action_space(self):
        return Box(low=0, high=1, shape=(2,))

    def step(self, action):
        return np.zeros(self.observation_space.shape), 1, False, {}

    def reset(self):
        return np.zeros(self.observation_space.shape)

def env_creator(env_config):
    return DummyEnv()

if __name__=='__main__':
    ray.init(object_store_memory=int(1e9), redis_max_memory=int(1e9), memory=int(1e9))

    alg_run = 'PPO'
    config = ppo.DEFAULT_CONFIG.copy()
    register_env('DummyEnv', env_creator)

    config['env'] = 'CrowdSim'
    config['num_workers'] = 2
    config['lr'] = tune.grid_search([1e-3, 1e-4, 1e-5, 1e-6])
    config['train_batch_size'] = 3000
    register_env('CrowdSim', env_creator)


    def trial_str_creator(trial):
        return "{}_{}".format(trial.trainable_name, trial.experiment_tag)


    exp_dict = {
        'name': 'Test',
        'run_or_experiment': alg_run,
        'checkpoint_freq': 1000,
        'stop': {
            'training_iteration': 10000
        },
        'config': config,
    }

    run_tune(**exp_dict, queue_trials=False)

`

**Comment by eugenevinitsky** [comment]
Checked it with a not multi-agent script just to confirm that this isn't confined to multi-agent envs.

**Comment by ericl** [comment]
Hmm that is surprising. Is it perhaps TensorFlow version difference?

Btw, how much memory does the script end up using at start / end? Another thing to try is "num_workers": 0.

**Comment by eugenevinitsky** [comment]
I have tensorflow 1.14.0 on both instances so I don't think it can be that.

**Comment by eugenevinitsky** [comment]
Just to add to the previous note, the memory usage increases until all system memory is consumed and the experiment fails.

**Comment by ericl** [comment]
At 1M steps, this is what I see:
```
1m steps
KiB Mem : 50360963+total, 48244320+free, 12867768 used,  8298668 buff/cache
 46302 ubuntu    20   0 5169364 718052 399916 S   1.7  0.1  60:38.89 ray::PPO.train(                                                                                                                        
 46294 ubuntu    20   0 5169364 715332 397156 S 337.7  0.1  61:49.23 ray::PPO.train(                                                                                                                        
 46297 ubuntu    20   0 5169316 714804 396632 S   1.7  0.1  60:30.65 ray::PPO.train(                                                                                                                        
 46339 ubuntu    20   0 5169372 708940 390696 S   1.7  0.1  61:22.45 ray::PPO.train(                                                                                                                        
 46321 ubuntu    20   0 3794980 659912 398508 S  95.0  0.1  63:13.06 ray::RolloutWor                                                                                                                        
 46352 ubuntu    20   0 3794472 659376 397996 R 108.6  0.1  63:03.54 ray::RolloutWor                                                                                                                        
 46545 ubuntu    20   0 3794900 659156 397328 R 108.9  0.1  63:19.78 ray::RolloutWor                                                                                                                        
 46317 ubuntu    20   0 3794904 658976 397208 S  95.7  0.1  62:59.83 ray::RolloutWor                                                                                                                        
 46350 ubuntu    20   0 3794908 658512 396776 R   2.3  0.1  63:06.96 ray::RolloutWor                                                                                                                        
 46357 ubuntu    20   0 3794524 657076 395692 R  15.2  0.1  62:56.36 ray::RolloutWor                                                                                                                        
 46316 ubuntu    20   0 3794540 652932 391548 R 108.9  0.1  64:40.55 ray::RolloutWor                                                                                                                        
 46343 ubuntu    20   0 3794952 652780 391324 S  95.7  0.1  62:09.68 ray::RolloutWor                                                                                                                        
```

It looks like about 12GB of ram is used only, which is about the same as on start.

**Comment by eugenevinitsky** [comment]
Ah I see, I'm on the nightly builds on an m4.10xlarge machine w/ tensorflow 1.14.0 and numpy 1.17.1. I figured if anything was the culprit it might be one of those? I'll try a clean AMI and see if the issue goes away.

**Comment by RolandZimmermann** [comment]
I have updated ray to the newest version along with a tensorflow 1.15 and numpy 1.17.5 the memory usage is lower, but the memorly leak still persists.

**Comment by eugenevinitsky** [comment]
Yeah, I'm still having the same error as well.

**Comment by eugenevinitsky** [comment]
@ericl could you share how you set up your AMI? We'd like to mimic it so we can debug where things start to go wrong.

**Comment by ericl** [comment]
I used the DLAMI, and installed the latest Ray wheel in the tensorflow_p36
env.

**Comment by eugenevinitsky** [comment]
Thank you! We'll update you when we find the change that causes it to break.

**Comment by eugenevinitsky** [comment]
@ericl I'm not sure what I'm doing differently than you, I installed Ray 0.7.7 on the DLAMI and get this graph:
<img width="272" alt="image" src="https://user-images.githubusercontent.com/7660397/71223331-7ca05380-2288-11ea-9e4d-34b1ffb28131.png">

I'll keep running it but in my experience it never levels off.

**Comment by eugenevinitsky** [comment]
@RolandZimmermann were you able to resolve this? Just like you're seeing for me the memory accumulates slower in TF 1.15.0

**Comment by eugenevinitsky** [comment]
This is the corresponding error btw: 
[error-2.txt](https://github.com/ray-project/ray/files/3986699/error-2.txt)

**Comment by RolandZimmermann** [comment]
@eugenevinitsky sadly not. I have experienced the same thing with TF 1.15. It indeed accumulates slower (TF 2.0 makes it even worse than before). But I couldn't figure out why thats the case. Just like in your experiments the memory usage goes up until it crashes.

**Comment by eugenevinitsky** [comment]
Update, this leak goes away with A3C. It seems to be restricted to PPO.

**Comment by eugenevinitsky** [comment]
@ericl are you sure you didn't see the leak? It's extremely slow, only a few GB after hundreds of iterations. Also, if it's helpful, confirmed that this occurs with A3C as well.

**Comment by ericl** [comment]
I'll be able to try this out again after the holidays. It's possible I
missed it if it's very slow.

One other thing to try debugging is stripping out parts of the loss, stats,
etc... Historically most of these leaks are due to adding more TF ops after
each loop.

**Comment by eugenevinitsky** [comment]
Okay, that's really helpful advice! I'll give it a go since it's really critical for us.

**Comment by eugenevinitsky** [comment]
Suspicious that it's the following lines in GAE:
`
@make_tf_callable(self.get_session())
            def value(ob, prev_action, prev_reward, *state):
                model_out, _ = self.model({
                    SampleBatch.CUR_OBS: tf.convert_to_tensor([ob]),
                    SampleBatch.PREV_ACTIONS: tf.convert_to_tensor(
                        [prev_action]),
                    SampleBatch.PREV_REWARDS: tf.convert_to_tensor(
                        [prev_reward]),
                    "is_training": tf.convert_to_tensor(False),
                }, [tf.convert_to_tensor([s]) for s in state],
                                          tf.convert_to_tensor([1]))
                return self.model.value_function()[0]
`
tf.convert_to_tensor appears to have a memory leak in 1.14.0. 
https://github.com/tensorflow/tensorflow/issues/31419

**Comment by eugenevinitsky** [comment]
Top growth is 1.14.0, bottom is 1.15.0. They start at different RAM percentages, but you can see that they grow at different speeds:
<img width="272" alt="image" src="https://user-images.githubusercontent.com/7660397/71412521-03a05380-261c-11ea-9879-1a6eea011644.png">

**Comment by ericl** [comment]
Interesting finds. So is the convert to tensor leak entirely fixed in 1.15?

**Comment by ericl** [comment]
Closing since this sounds fixed with TF1.15.
